### PR TITLE
ica: wrong handshake flow tests

### PR DESCRIPTION
**Pull request by colin-axner**
## Description

closes: #550 
closes: #522 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [x] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: modules/apps/27-interchain-accounts/interchain_accounts_test.go
Comment (seantking) [inline modules/apps/27-interchain-accounts/interchain_accounts_test.go:21]: Am I correct in saying this only fails because we don't have a Path setup for Chain B -> Chain A?

File: modules/apps/27-interchain-accounts/interchain_accounts_test.go
Comment (seantking) [inline modules/apps/27-interchain-accounts/interchain_accounts_test.go:39]: Question: Why do we need to make use of this util here? I run into client issues sometimes when writing tests so would like to better understand. :handshake: 

Is this essentially just updating Chain A with the latest state from chain B?

File: modules/apps/27-interchain-accounts/interchain_accounts_test.go
Comment (seantking) [inline modules/apps/27-interchain-accounts/interchain_accounts_test.go:64]: Is the idea that by doing this a relayer will try and call `OnChanOpenAck` on the host chain as the next step? What would be the benefit of this to the malicious chain A? 

Just trying to better understand what we are testing here (other than the sanity check) :)

**Review by seantking — APPROVED** [review]

File: modules/apps/27-interchain-accounts/interchain_accounts_test.go
Comment (colin-axner) [reply modules/apps/27-interchain-accounts/interchain_accounts_test.go:21]: This fails because OpenInit is called on the host chain port ID

File: modules/apps/27-interchain-accounts/interchain_accounts_test.go
Comment (colin-axner) [reply modules/apps/27-interchain-accounts/interchain_accounts_test.go:39]: MsgChannelOpenTry is being delivered to chainA with proofs from chainB

We made state changes on chainB so in order for those proofs to pass on chainA, we need to update the light client on chainA to include a consensus state which can verify the new state changes

File: modules/apps/27-interchain-accounts/interchain_accounts_test.go
Comment (colin-axner) [reply modules/apps/27-interchain-accounts/interchain_accounts_test.go:64]: > Is the idea that by doing this a relayer will try and call `OnChanOpenAck` on the host chain as the next step?

The idea is that the relayer could call OnChanOpenAck on the controller chain using the host port ID. 

>  What would be the benefit of this to the malicious chain A?

The host portID would be marked as an active channel. I'm not sure if there is any benefit. This is mostly just a sanity check, the existing logic makes it difficult to confuse a host or controller chain

**Comment by colin-axner** [comment]
> Nice! If you want to go ahead and merge this I can try port these tests to #541. Or if you think it's easier to wait until after we can do that too :)

I will wait!

**Comment by colin-axner** [comment]
Updated to add invalid flow tests to both controller/host where applicable. Only Ack/Timeout needs code coverage in controller now

File: modules/apps/27-interchain-accounts/controller/ibc_module_test.go
Comment (damiannolan) [inline modules/apps/27-interchain-accounts/controller/ibc_module_test.go:213]: So this `QueryProof` on the testchain just calls into ABCI for a proof that this channelKey is set in the store? 
Do I understand that correctly? 

What happens if it doesn't exist? Do you get an empty `[]byte` and height of -1 or something? 

Just asking out of my own curiosity :)

**Review by damiannolan — APPROVED** [review]
Awesome! Really appreciate you holding off on this for the submodules too, many thanks 🙌

File: modules/apps/27-interchain-accounts/controller/ibc_module_test.go
Comment (colin-axner) [reply modules/apps/27-interchain-accounts/controller/ibc_module_test.go:213]: > So this QueryProof on the testchain just calls into ABCI for a proof that this channelKey is set in the store?
Do I understand that correctly?

Correct! That the key is set in store with the expected value

> What happens if it doesn't exist? Do you get an empty []byte and height of -1 or something?

It returns a non existence proof

**Comment by damiannolan** [comment]
It looks like there's a failing test in the CI actually.

File: modules/apps/27-interchain-accounts/host/ibc_module_test.go
Comment (damiannolan) [inline modules/apps/27-interchain-accounts/host/ibc_module_test.go:487]: I think this fixes the failing test

```suggestion
			module, _, err := suite.chainA.App.GetIBCKeeper().PortKeeper.LookupModuleByPort(suite.chainA.GetContext(), path.EndpointB.ChannelConfig.PortID)
```

File: modules/apps/27-interchain-accounts/controller/ibc_module_test.go
Comment (damiannolan) [reply modules/apps/27-interchain-accounts/controller/ibc_module_test.go:213]: Ah okay, great! So if you were to use the non existence proof then as argument to some other call, an error would eventually be propagated back to you I assume?

Comment (colin-axner) [reply modules/apps/27-interchain-accounts/controller/ibc_module_test.go:213]: correct, the non existence proof should be rejected by the tendermint client

**Comment by colin-axner** [comment]
Clicking auto-merge. The only non test code change has been there since the beginning (2 approvals). The test code changes seem fine to be merged with 1 approval

**Pull request merged**
